File: 49. Group Anagrams/main.py
from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        sorted_strs = [''.join(sorted(s)) for s in strs]
        mapping = {}
        for i in range(len(strs)):
            if mapping.get(sorted_strs[i]):
                mapping[sorted_strs[i]].append(strs[i])
            else:
                mapping[sorted_strs[i]] = [strs[i]]
        return list(mapping.values())

s = Solution()
print(s.groupAnagrams(["eat","tea","tan","ate","nat","bat"])) # [["bat"],["nat","tan"],["ate","eat","tea"]]
print(s.groupAnagrams([""])) # [[""]]
print(s.groupAnagrams(["a"])) # [["a"]]

File: 49. Group Anagrams/main.js
var groupAnagrams = function (strs) {
    let result = [];
    let sorted = strs.map(s => s.split('').sort().join(''));
    let map = {};
    for (let i = 0; i < sorted.length; i++) {
        if (!map[sorted[i]]) map[sorted[i]] = [strs[i]];
        else {
            map[sorted[i]].push(strs[i]);
        }
    }
    return Object.values(map);
};

// Test Case 1
let input = ["eat", "tea", "tan", "ate", "nat", "bat"];
let output = groupAnagrams(input);
console.log(output);
// Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

//Test Case 2
input = ["listen", "silent", "below", "elbow", "state", "taste"];
output = groupAnagrams(input);
console.log(output);
// Output: [["listen","silent"],["below","elbow"],["state","taste"]]
